File: tools/tests/test_site_js.js
/* The browser half, held to the same answers as the device.
 *
 * site/app.js is loaded for real and driven against the baked year a static build
 * produces. What is being checked is the claim the whole design rests on: the pages carry
 * no prayer rules, they only look up which interval the clock is in - so for every second
 * of a day, the colour the page would paint must be the colour period_state chose.
 *
 * The Python side of that comparison is written out by tools/tests/test_site_js.py, which
 * is what runs this file. Run that, not this.
 */

const fs = require("fs");
const path = require("path");
const vm = require("vm");

const [, , sitePath, yearPath, expectedPath] = process.argv;

/* app.js declares `const Site = ...` at top level and nothing else, so it is evaluated in
 * a context with a `window` and its export picked back up. No DOM is needed: everything
 * being tested here is arithmetic over the payload. */
const context = { window: {}, console, fetch: undefined };
vm.createContext(context);
/* A top-level `const` in a vm context lands in the declarative record, not on the context
   object, so it is picked up as the completion value instead. */
const source = fs.readFileSync(path.join(sitePath, "static", "app.js"), "utf8");
const Site = vm.runInContext(source + "\n;Site;", context, { filename: "app.js" });

const year = JSON.parse(fs.readFileSync(yearPath, "utf8"));
const expected = JSON.parse(fs.readFileSync(expectedPath, "utf8"));

let checks = 0;
const failures = [];

function check(label, got, want) {
    checks += 1;
    if (got !== want) failures.push(`  ${label}\n    got  ${got}\n    want ${want}`);
}

/* 1. parseLocal must read the server's timestamps as local time. Date's own parser
   treats "2026-03-10" as UTC, which would shift every boundary by the timezone offset -
   the bug this function exists to avoid. */
const parsed = Site.parseLocal("2026-03-10T05:12:00");
check("parseLocal year", parsed.getFullYear(), 2026);
check("parseLocal month", parsed.getMonth(), 2);
check("parseLocal day", parsed.getDate(), 10);
check("parseLocal hour", parsed.getHours(), 5);
check("parseLocal minute", parsed.getMinutes(), 12);
check("a bare date is still local midnight",
      Site.parseLocal("2026-03-10").getHours(), 0);
check("nonsense is refused", Site.parseLocal("not a date"), null);
check("and so is an empty string", Site.parseLocal(""), null);

/* 2. clock12 must read the way clock_12h does on the device. */
for (const [iso, want] of Object.entries(expected.clocks)) {
    check(`clock12 ${iso}`, Site.clock12(Site.parseLocal(iso)), want.trim());
}

/* 3. The whole point. For every second of the sample day, the span the page would pick
   must carry the colour and the makrooh flag Python chose. */
let mismatches = 0;
let compared = 0;
for (const sample of expected.samples) {
    const periods = year.days[sample.key];
    if (!periods) {
        failures.push(`  no day ${sample.key} in the baked year`);
        continue;
    }
    const period = periods.find((p) => p.name === sample.name);
    if (!period) {
        failures.push(`  no period ${sample.name} on ${sample.key}`);
        continue;
    }
    const span = Site.spanAt(period, Site.parseLocal(sample.at));
    compared += 1;
    if (!span || span.color !== sample.color || span.makrooh !== sample.makrooh) {
        mismatches += 1;
        if (mismatches <= 5) {
            failures.push(`  ${sample.name} at ${sample.at}: page says ` +
                `${span ? span.color + "/" + span.makrooh : "nothing"}, ` +
                `device says ${sample.color}/${sample.makrooh}`);
        }
    }
}
checks += 1;
if (mismatches) failures.push(`  ${mismatches} of ${compared} samples disagree`);

/* 4. runningPeriod must pick the period the clock is inside, and nothing when it is not
   inside any - which is what the countdown page uses to decide to re-read the day. */
for (const sample of expected.running) {
    const periods = year.days[sample.key];
    const found = Site.runningPeriod({ periods }, Site.parseLocal(sample.at));
    check(`runningPeriod ${sample.at}`, found ? found.name : null, sample.name);
}

/* 5. The premise the countdown's post-midnight fallback rests on: in a baked year, the
   hours after midnight are covered by the previous day's Isha and by nothing on the
   day itself. This is what showed up in the field as a counter reading "--:--". */
{
    const sample = expected.smallHours;
    const todayPeriods = year.days[sample.todayKey];
    const beforePeriods = year.days[sample.yesterdayKey];
    const at = Site.parseLocal(sample.at);

    check("nothing on today's table covers 01:00",
          Site.runningPeriod({ periods: todayPeriods }, at), null);
    const found = Site.runningPeriod({ periods: beforePeriods }, at);
    check("the evening before's table does", found ? found.name : null, "العشاء");
}

/* 6. isoDate must round-trip the key the baked year is keyed by. */
const d = new Date(2026, 2, 10, 13, 0, 0);
check("isoDate", Site.isoDate(d), "2026-03-10");

console.log("");
if (failures.length) {
    console.log(`FAILED (${failures.length} of ${checks} checks)`);
    console.log(failures.join("\n"));
    process.exit(1);
}
console.log(`ALL PASS (${checks} checks, ${compared} second-by-second samples)`);

File: tools/tests/test_site_js.py
#!/usr/bin/env python3
"""Does the browser paint what the device decided?

Builds a real static site, works out the right answer in Python with the same
prayer_logic the touch screen uses, and hands both to node so site/app.js can be held to
it. That closes the one gap the Python tests cannot reach: the pages are supposed to carry
no prayer rules at all, and this is what proves the lookup they do instead agrees.

Skipped, not failed, where there is no node - the device does not need one, and this is a
check on the front end rather than on anything that ships.
"""

import json
import os
import shutil
import subprocess
import sys
import tempfile
from datetime import date as date_cls, datetime, timedelta

REPO = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
VARIANT = "scheduler-official-touch-screen-with-raspberry-pi-4"
APPLICATIONS = os.path.join(REPO, VARIANT, "applications")
sys.path.insert(0, APPLICATIONS)

from shared import prayer_logic  # noqa: E402

if shutil.which("node") is None:
    print("SKIPPED - no node on this machine")
    sys.exit(0)

YEAR = 2026
SAMPLE = date_cls(YEAR, 3, 10)

ROOT = tempfile.mkdtemp(prefix="site-js-test-")
SCHEDULER = os.path.join(ROOT, "scheduler")
os.makedirs(os.path.join(SCHEDULER, "config", "fonts", "arabic-fonts"))

ROWS = []
day = date_cls(YEAR, 1, 1)
while day.year == YEAR:
    ROWS.append({"Month": day.month, "Day": day.day, "Fajr": "05:00",
                 "Sunrise": "06:00", "Dhuhr": "12:00", "Asr": "15:00",
                 "Maghrib": "18:00", "Isha": "20:00"})
    day += timedelta(days=1)
MAP = os.path.join(SCHEDULER, "config", "prayer_times_map.py")
with open(MAP, "w", encoding="utf-8") as handle:
    handle.write("prayerTimes = " + repr(ROWS))

OUT = os.path.join(ROOT, "site")
build = subprocess.run([sys.executable, os.path.join(REPO, "tools", "build_static_site.py"),
                        "--scheduler-dir", SCHEDULER, "--out", OUT, "--year", str(YEAR)],
                       capture_output=True, text=True)
if build.returncode != 0:
    print("FAILED - the static build did not run")
    print(build.stdout, build.stderr)
    sys.exit(1)

prayer_logic.PRAYER_MAP_FILE = MAP
assert prayer_logic.load_prayer_times()

noon = datetime(SAMPLE.year, SAMPLE.month, SAMPLE.day, 13, 0)
rows = prayer_logic.daily_rows(SAMPLE, noon, False)
key = prayer_logic.key_for_date(SAMPLE)

# Every 30 seconds of every period, with the answer the device would give.
samples = []
for row in rows:
    if not (row["start"] and row["end"]):
        continue
    moment = row["start"]
    while moment < row["end"]:
        color, makrooh = prayer_logic.period_state(row["name"], row["start"],
                                                   row["end"], moment)
        samples.append({"key": key, "name": row["name"],
                        "at": moment.isoformat(timespec="seconds"),
                        "color": color, "makrooh": makrooh})
        moment += timedelta(seconds=30)

clocks = {}
for hour, minute in ((5, 12), (12, 18), (0, 5), (20, 0), (13, 45), (11, 59)):
    when = datetime(SAMPLE.year, SAMPLE.month, SAMPLE.day, hour, minute)
    clocks[when.isoformat(timespec="seconds")] = prayer_logic.clock_12h(when)

# Which period is in force at a few instants, including one inside no period at all.
running = []
for hour, minute in ((5, 30), (6, 10), (9, 0), (13, 0), (16, 0), (19, 0), (21, 0)):
    when = datetime(SAMPLE.year, SAMPLE.month, SAMPLE.day, hour, minute)
    found = None
    for row in rows:
        if row["start"] and row["end"] and row["start"] <= when < row["end"]:
            found = row["name"]
            break
    running.append({"key": key, "at": when.isoformat(timespec="seconds"), "name": found})

# 01:00 on the day after the sample: in a baked year that moment is inside the previous
# day's Isha and inside nothing at all on the day itself, which is why the countdown page
# looks back a day when it finds nothing running.
after_midnight = datetime(SAMPLE.year, SAMPLE.month, SAMPLE.day, 1, 0) + timedelta(days=1)
small_hours = {
    "at": after_midnight.isoformat(timespec="seconds"),
    "todayKey": prayer_logic.key_for_date(after_midnight.date()),
    "yesterdayKey": prayer_logic.key_for_date(after_midnight.date() - timedelta(days=1)),
}

EXPECTED = os.path.join(ROOT, "expected.json")
with open(EXPECTED, "w", encoding="utf-8") as handle:
    json.dump({"samples": samples, "clocks": clocks, "running": running,
               "smallHours": small_hours}, handle, ensure_ascii=False)

result = subprocess.run(["node", os.path.join(REPO, "tools", "tests", "test_site_js.js"),
                         OUT, os.path.join(OUT, "data", "year.json"), EXPECTED],
                        capture_output=True, text=True)
print(result.stdout.strip())
if result.stderr.strip():
    print(result.stderr.strip())

shutil.rmtree(ROOT, ignore_errors=True)
sys.exit(result.returncode)
